src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -244,36 +265,90 @@ QueryPlanPtr MergeTreeDataSelectExecutor::readFromParts(
+            auto ignored_uuids_map = query_context.getIgnoredUUIDs()->getMap();
+            std::set<UUID> temp_uuids;
 
-            if (part->isEmpty())
-                continue;
-
-            if (minmax_idx_condition && !minmax_idx_condition->checkInHyperrectangle(
-                    part->minmax_idx.hyperrectangle, data.minmax_idx_column_types).can_be_true)
-                continue;
+            auto prev_parts = parts;
+            parts.clear();
 
-            if (partition_pruner)
+            for (const auto & part : prev_parts)
             {
-                if (partition_pruner->canBePruned(part))
+                if (part_values.find(part->name) == part_values.end())
+                    continue;
+
+                if (part->isEmpty())
                     continue;
+
+                if (minmax_idx_condition
+                    && !minmax_idx_condition->checkInHyperrectangle(part->minmax_idx.hyperrectangle, data.minmax_idx_column_types)
+                            .can_be_true)
+                    continue;
+
+                if (partition_pruner)
+                {
+                    if (partition_pruner->canBePruned(part))
+                        continue;
+                }
+
+                if (max_block_numbers_to_read)
+                {
+                    auto blocks_iterator = max_block_numbers_to_read->find(part->info.partition_id);
+                    if (blocks_iterator == max_block_numbers_to_read->end() || part->info.max_block > blocks_iterator->second)
+                        continue;
+                }
+
+                /// populate UUIDs and exclude ignored parts if enabled
+                if (query_context.getSettingsRef().allow_experimental_query_deduplication && part->uuid != UUIDHelpers::Nil)
+                {
+                    /// Skip the part if its uuid is meant to be excluded
+                    if unlikely (ignored_uuids_map.find(part->uuid) != ignored_uuids_map.end())
+                        continue;
+
+                    auto result = temp_uuids.insert(part->uuid);
+                    /// found a part with the same uuid one the same replica, ignore it
+                    if (!result.second)
+                        continue;
+                }
+
+                parts.push_back(part);
             }
 
-            if (max_block_numbers_to_read)
+            if unlikely (!temp_uuids.empty())
             {
-                auto blocks_iterator = max_block_numbers_to_read->find(part->info.partition_id);
-                if (blocks_iterator == max_block_numbers_to_read->end() || part->info.max_block > blocks_iterator->second)
-                    continue;
+                auto duplicates = query_context.getUUIDs()->add(std::vector<UUID>{temp_uuids.begin(), temp_uuids.end()});
+                if (!duplicates.empty())
+                {
+                    /// if any duplicates appeared during the first pass, adding them to the exclusion, so they will be skipped on second pass
```

Comment:
When this can happen?

Reply:
Need to add this into the comments, I still don't undestand.